File: fugaku_in_practice/hands_on/sample_code/01_processor-core/01_polynomial/fortran_AssumedShapeArray/mykernel.f90
! Copyright 2024 Research Organization for Information Science and Technology
module mykernel
  use mytype, only: DP
  implicit none
contains
  subroutine poly1(x, c1, c2)
     real(kind=DP), dimension(:), contiguous, intent(inout) :: x
     real(kind=DP), dimension(:), contiguous, intent(in) :: c1
     real(kind=DP), dimension(:), contiguous, intent(in) :: c2
    
     x(:) = c1(:) + x(:)*c2(:)
  end subroutine poly1

  subroutine poly4(x, c1, c2)
     real(kind=DP), dimension(:), contiguous, intent(inout) :: x
     real(kind=DP), dimension(:), contiguous, intent(in) :: c1
     real(kind=DP), dimension(:), contiguous, intent(in) :: c2
    
     x(:) = c1(:) + x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*c1(:))))
  end subroutine poly4

  subroutine poly8(x, c1, c2)
     real(kind=DP), dimension(:), contiguous, intent(inout) :: x
     real(kind=DP), dimension(:), contiguous, intent(in) :: c1
     real(kind=DP), dimension(:), contiguous, intent(in) :: c2
    
     x(:) = c1(:)                                           &
           +x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*(c1(:) &
           +x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*c1(:)  &
           )))))))
  end subroutine poly8

  subroutine poly16(x, c1, c2)
     real(kind=DP), dimension(:), contiguous, intent(inout) :: x
     real(kind=DP), dimension(:), contiguous, intent(in) :: c1
     real(kind=DP), dimension(:), contiguous, intent(in) :: c2
    
     x(:) = c1(:)                                           &
           +x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*(c1(:) &
           +x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*(c1(:) &
           +x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*(c1(:) &
           +x(:)*(c2(:)+x(:)*(c1(:)+x(:)*(c2(:)+x(:)*c1(:)  &
           )))))))))))))))
  end subroutine poly16
end module mykernel

File: fugaku_in_practice/hands_on/sample_code/01_processor-core/01_polynomial/cpp_ExpressionTemplate/mykernel.cpp
// Copyright 2024 Research Organization for Information Science and Technology
#include "mykernel.h"

void poly1 (vector_et<ARRAY_SIZE> &x ,
            vector_et<ARRAY_SIZE> &c0, vector_et<ARRAY_SIZE> &c1)
{
#ifdef __ICC_FORCEINLINE__
#pragma forceinline recursive // for icc
#endif // __ICC_FORCEINLINE__
   x = c0+x*c1;
}

void poly4 (vector_et<ARRAY_SIZE> &x ,
            vector_et<ARRAY_SIZE> &c0, vector_et<ARRAY_SIZE> &c1)
{
#ifdef __ICC_FORCEINLINE__
#pragma forceinline recursive // for icc
#endif // __ICC_FORCEINLINE__
   x = c0 
         +x*(c1 +x*(c0 +x*(c1 +x*c0))) ;
}

void poly8 (vector_et<ARRAY_SIZE> &x ,
            vector_et<ARRAY_SIZE> &c0, vector_et<ARRAY_SIZE> &c1)
{
#ifdef __ICC_FORCEINLINE__
#pragma forceinline recursive // for icc
#endif // __ICC_FORCEINLINE__
    x = c0 
          +x*(c1 +x*(c0 +x*(c1 +x*(c0 
	  +x*(c1 +x*(c0 +x*(c1 +x*c0 
          ))))))) ;
}

void poly16 (vector_et<ARRAY_SIZE> &x ,
            vector_et<ARRAY_SIZE> &c0, vector_et<ARRAY_SIZE> &c1)
{
#ifdef __ICC_FORCEINLINE__
#pragma forceinline recursive // for icc
#endif // __ICC_FORCEINLINE__
    x = c0 
          +x*(c1 +x*(c0 +x*(c1 +x*(c0
	  +x*(c1 +x*(c0 +x*(c1 +x*(c0 
	  +x*(c1 +x*(c0 +x*(c1 +x*(c0 
	  +x*(c1 +x*(c0 +x*(c1 +x*c0 
          ))))))))))))))) ;
}
